File: 494. Target Sum/recursive.py
class Solution:
    def findTargetSumWays(self, nums: List[int], target: int) -> int:         # type: ignore
        def dp(i, total, memo):
            if (i, total) in memo:
                return memo[i, total]
            if i == len(nums) and total == target:
                return 1
            if i == len(nums) and total != target:
                return 0             
            memo[i, total] = dp(i + 1, total + nums[i], memo) + dp(i + 1, total - nums[i], memo)
            return memo[i, total]
        return dp(0, 0, {})

File: 494. Target Sum/recursive.rs
use std::collections::HashMap;
impl Solution {
    pub fn find_target_sum_ways(nums: Vec<i32>, target: i32) -> i32 {
        return Self::ways(0, 0, &nums, &target, &mut HashMap::new())
    }
    pub fn ways(i: i32, total: i32, nums: &Vec<i32>, target: &i32, memo: &mut HashMap<(i32, i32), i32>)->i32{
        if let Some(&cached) = memo.get(&(i, total)){
            return cached;
        }
        if i as usize == nums.len() && total == *target{
            return 1;
        }
        if i as usize == nums.len() && total != *target{
            return 0;
        }
        let res = Self::ways(i + 1, total + nums[i as usize], nums, target, memo) + 
                    Self::ways(i + 1, total - nums[i as usize], nums, target, memo);
        memo.insert((i, total), res);
        return res;
    }
}